File: reference-implementation/go/futures.go
package main

import (
	"context"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// Mock chain is a fixed snapshot as of 2026-11-06 (42 days before ESZ26 expiry):
// ESU26 has expired, ESZ26 is front month, ESH27 is next out.
const (
	futuresRootID       = "APEX:FUT:ES"
	futuresExpiredID    = "APEX:FUT:ESU26"
	futuresFrontMonthID = "APEX:FUT:ESZ26"
	futuresNextMonthID  = "APEX:FUT:ESH27"
)

func registerFuturesToolsWithState(s *server.MCPServer, st *referenceState) {
	// apex.futures.contract_chain
	s.AddTool(
		mcp.NewTool("apex.futures.contract_chain",
			mcp.WithDescription("List dated contracts for a futures contract root with expirations and liquidity. Reference implementation serves the E-mini S&P 500 chain."),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
			mcp.WithString("root", mcp.Required(), mcp.Description("APEX contract root ID (e.g. APEX:FUT:ES)")),
			mcp.WithBoolean("include_expired", mcp.Description("Include expired contracts (default: false)")),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			root := strParam(args, "root", "")
			includeExpired, _ := args["include_expired"].(bool)
			if root == "" {
				return jsonResult(apexError("APEX_4011", "validation", "root is required"))
			}
			if root != futuresRootID {
				return jsonResult(apexError("APEX_4010", "validation", "Unknown instrument"))
			}

			contracts := []map[string]any{}
			if includeExpired {
				contracts = append(contracts, map[string]any{
					"instrument_id":     futuresExpiredID,
					"contract_month":    "2026-09",
					"expiration_date":   "2026-09-18",
					"first_notice_date": nil,
					"settlement_type":   "cash",
					"is_front_month":    false,
					"volume":            0,
					"open_interest":     0,
					"status":            "inactive",
				})
			}
			contracts = append(contracts,
				map[string]any{
					"instrument_id":     futuresFrontMonthID,
					"contract_month":    "2026-12",
					"expiration_date":   "2026-12-18",
					"first_notice_date": nil,
					"settlement_type":   "cash",
					"is_front_month":    true,
					"volume":            1250000,
					"open_interest":     2100000,
					"status":            "active",
				},
				map[string]any{
					"instrument_id":     futuresNextMonthID,
					"contract_month":    "2027-03",
					"expiration_date":   "2027-03-19",
					"first_notice_date": nil,
					"settlement_type":   "cash",
					"is_front_month":    false,
					"volume":            41000,
					"open_interest":     98000,
					"status":            "active",
				},
			)

			return jsonResult(map[string]any{
				"root":      futuresRootID,
				"contracts": contracts,
			})
		},
	)

	// apex.futures.margin_schedule
	s.AddTool(
		mcp.NewTool("apex.futures.margin_schedule",
			mcp.WithDescription("Per-contract margin requirements: exchange overnight margins and broker intraday margins. Reference implementation serves the ESZ26 schedule."),
			mcp.WithReadOnlyHintAnnotation(true),
			mcp.WithDestructiveHintAnnotation(false),
			mcp.WithIdempotentHintAnnotation(true),
			mcp.WithString("account_id", mcp.Required(), mcp.Description("Trading account ID")),
			mcp.WithString("instrument_id", mcp.Description("Filter by APEX canonical instrument ID (e.g. APEX:FUT:ESZ26)")),
		),
		func(_ context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			args := request.GetArguments()
			accountID := strParam(args, "account_id", "")
			instrumentID := strParam(args, "instrument_id", "")

			if accountID == "" {
				return jsonResult(apexError("APEX_4011", "validation", "account_id is required"))
			}
			if instrumentID != "" && instrumentID != futuresFrontMonthID {
				return jsonResult(apexError("APEX_4010", "validation", "Unknown instrument"))
			}

			return jsonResult(map[string]any{
				"margins": []map[string]any{
					{
						"instrument_id":      futuresFrontMonthID,
						"currency":           "USD",
						"initial_margin":     "15500.00",
						"maintenance_margin": "14000.00",
						"day_trading_margin": "500.00",
						"day_trading_hours": []map[string]any{
							{"day": "monday", "from": "08:30", "to": "15:45", "timezone": "America/Chicago"},
						},
						"as_of": nowISO(),
					},
				},
			})
		},
	)
}

File: reference-implementation/typescript/src/tools/futures.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import type { ReferenceTradingState } from "../lib/resources.js";
import { apexError, nowIso } from "../lib/helpers.js";

// Mock chain is a fixed snapshot as of 2026-11-06 (42 days before ESZ26 expiry):
// ESU26 has expired, ESZ26 is front month, ESH27 is next out.
const FUTURES_ROOT_ID = "APEX:FUT:ES";
const FUTURES_EXPIRED_ID = "APEX:FUT:ESU26";
const FUTURES_FRONT_MONTH_ID = "APEX:FUT:ESZ26";
const FUTURES_NEXT_MONTH_ID = "APEX:FUT:ESH27";

export function registerFuturesTools(server: McpServer, state: ReferenceTradingState): void {
  /* ------------------------------------------------------------------ */
  /*  apex.futures.contract_chain                                        */
  /* ------------------------------------------------------------------ */

  server.registerTool(
    "apex.futures.contract_chain",
    {
      description:
        "List dated contracts for a futures contract root with expirations and liquidity. Reference implementation serves the E-mini S&P 500 chain.",
      inputSchema: {
        root: z.string().describe("APEX contract root ID (e.g. APEX:FUT:ES)"),
        include_expired: z.boolean().optional().describe("Include expired contracts (default: false)"),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async ({ root, include_expired }) => {
      if (!root) {
        return {
          structuredContent: apexError("APEX_4011", "validation", "root is required"),
          content: [],
        };
      }
      if (root !== FUTURES_ROOT_ID) {
        return {
          structuredContent: apexError("APEX_4010", "validation", "Unknown instrument"),
          content: [],
        };
      }

      const expired = include_expired
        ? [
            {
              instrument_id: FUTURES_EXPIRED_ID,
              contract_month: "2026-09",
              expiration_date: "2026-09-18",
              first_notice_date: null,
              settlement_type: "cash",
              is_front_month: false,
              volume: 0,
              open_interest: 0,
              status: "inactive",
            },
          ]
        : [];

      return {
        structuredContent: {
          root: FUTURES_ROOT_ID,
          contracts: [
            ...expired,
            {
              instrument_id: FUTURES_FRONT_MONTH_ID,
              contract_month: "2026-12",
              expiration_date: "2026-12-18",
              first_notice_date: null,
              settlement_type: "cash",
              is_front_month: true,
              volume: 1250000,
              open_interest: 2100000,
              status: "active",
            },
            {
              instrument_id: FUTURES_NEXT_MONTH_ID,
              contract_month: "2027-03",
              expiration_date: "2027-03-19",
              first_notice_date: null,
              settlement_type: "cash",
              is_front_month: false,
              volume: 41000,
              open_interest: 98000,
              status: "active",
            },
          ],
        },
        content: [],
      };
    },
  );

  /* ------------------------------------------------------------------ */
  /*  apex.futures.margin_schedule                                       */
  /* ------------------------------------------------------------------ */

  server.registerTool(
    "apex.futures.margin_schedule",
    {
      description:
        "Per-contract margin requirements: exchange overnight margins and broker intraday margins. Reference implementation serves the ESZ26 schedule.",
      inputSchema: {
        account_id: z.string().describe("Trading account ID"),
        instrument_id: z
          .string()
          .optional()
          .describe("Filter by APEX canonical instrument ID (e.g. APEX:FUT:ESZ26)"),
      },
      annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    },
    async ({ account_id, instrument_id }) => {
      if (!account_id) {
        return {
          structuredContent: apexError("APEX_4011", "validation", "account_id is required"),
          content: [],
        };
      }
      if (instrument_id && instrument_id !== FUTURES_FRONT_MONTH_ID) {
        return {
          structuredContent: apexError("APEX_4010", "validation", "Unknown instrument"),
          content: [],
        };
      }

      return {
        structuredContent: {
          margins: [
            {
              instrument_id: FUTURES_FRONT_MONTH_ID,
              currency: "USD",
              initial_margin: "15500.00",
              maintenance_margin: "14000.00",
              day_trading_margin: "500.00",
              day_trading_hours: [
                { day: "monday", from: "08:30", to: "15:45", timezone: "America/Chicago" },
              ],
              as_of: nowIso(),
            },
          ],
        },
        content: [],
      };
    },
  );
}
